bundle/src/main/java/com/adobe/acs/commons/remoteassets/impl/RemoteAssetsNodeSyncImpl.java

```diff
@@ -0,0 +1,663 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package com.adobe.acs.commons.remoteassets.impl;
+
+import com.adobe.acs.commons.assets.FileExtensionMimeTypeConstants;
+import com.adobe.acs.commons.remoteassets.RemoteAssetsNodeSync;
+import com.adobe.acs.commons.remoteassets.RemoteAssetsConfig;
+import com.day.cq.commons.jcr.JcrUtil;
+import com.day.cq.dam.api.DamConstants;
+import com.day.cq.tagging.Tag;
+import com.day.cq.tagging.TagManager;
+import com.day.cq.wcm.api.NameConstants;
+import org.apache.commons.io.FilenameUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.felix.scr.annotations.Activate;
+import org.apache.felix.scr.annotations.Component;
+import org.apache.felix.scr.annotations.Deactivate;
+import org.apache.felix.scr.annotations.Reference;
+import org.apache.felix.scr.annotations.Service;
+import org.apache.jackrabbit.commons.JcrUtils;
+import org.apache.jackrabbit.oak.spi.security.authorization.accesscontrol.AccessControlConstants;
+import org.apache.jackrabbit.vault.util.JcrConstants;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.api.resource.ResourceResolverFactory;
+import org.apache.sling.commons.classloader.DynamicClassLoaderManager;
+import org.apache.sling.commons.json.JSONArray;
+import org.apache.sling.commons.json.JSONException;
+import org.apache.sling.commons.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.jcr.Node;
+import javax.jcr.RepositoryException;
+import javax.jcr.Session;
+import javax.jcr.Value;
+import javax.jcr.ValueFactory;
+import java.io.BufferedReader;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.math.BigDecimal;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Calendar;
+import java.util.GregorianCalendar;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+/**
+ * Service to sync node tree from supplied felix configuration. Implements {@link RemoteAssetsNodeSync}.
+ */
+@Component(
```

Comment:
I think the big potential room for improvement would be using Session.importXML and pulling the asset down AS XML, and letting the JCR APIs sort out all the data transformations. It would cut this method down to a handful of lines.

All that JSON / property setting is scary :)

Reply:
Ha ok...so you're recommending that we use JSON calls to traverse the tree, but then when we have a single asset call for the XML and sync the asset in that way?

Couple questions on that...
1) It is recommended that default XML renderer be turned off on production instances - I believe this would require the remote server to have XML renderer turned on - not sure if this is a big deal or not - thoughts?
2) We'd still need to do some contortions to add in the temporary binary files when creating the nodes.

I agree that this code is somewhat scary, just want to make sure there is a strong and good reason to change it before we do...

Reply:
Saw your comment below about including the binary data.  We actually dont want the binary data included in this request.  This request is getting all of the node structure *except for* the binary files - that makes this sync both fast (at least fast-er) and SMALL (this is one of the main tenets of Remote Assets).

If we switch to XML, and there's no way to exclude the jcr:data properties from the fetched content, I think this could dramatically increase the cost of this sync on both AEM servers involved.